File: PyBag.m
classdef PyBag
    %PYBAG A class for reading data from ROS bags using the Python API.
    %
    % This class is only compatible with Matlab version 2014b or later.
    
    properties
        path % path to the bag
        time_begin % timestamp of first message in the bag
        time_end % timestamp of last message in the bag
    end
    
    properties (Hidden)
        bag
    end
    
    methods(Static)
        function [obj] = load(path)
            % Construct a bag using path.
            %
            % Function exists because it autocompletes paths when you hit
            % tab.
            obj =  PyBag(path);
        end
    end
    
    methods
        function obj = PyBag(path)
            if verLessThan('matlab', '8.4')
                error('Python only supported since R2014b (Version 8.4)')
            end
            
            if ~exist(path, 'file')
                error('The input file ''%s'' does not exist', path)
            end
            
            obj.path = path;
            obj.bag = py.pybag.PyBag(path);
            obj.time_begin = obj.bag.start_time();
            obj.time_end = obj.bag.end_time();
        end
        
        function info = info(obj)
            info = python2matlab(obj.bag.info());
        end
        
        function [topics, types] = topics(obj)
            topics = python2matlab(obj.bag.topics());
            if nargout > 1
                types = python2matlab(obj.bag.types());
            end
        end
        
        function data = data(obj)
            data = python2matlab(obj.bag.data());
        end
        
    end
    
end

File: pybag.py
#!/usr/bin/env python
"""Python class for reading data from rosbag files into Matlab"""

__author__ = "Daniel Koch <daniel.koch@byu.edu>"
__version__ = 0.1

import rosbag
import yaml

class PyBag:
    """Class for extracting data from rosbag files"""

    def __init__(self, path):
        self.bag = rosbag.Bag(path, 'r')

    def __exit(self):
        self.bag.close()

    def info(self):
        return yaml.load(self.bag._get_yaml_info())

    def topics(self):
        """Return a list of all topics in the bag"""
        return self.bag.get_type_and_topic_info()[1].keys()

    def types(self):
        """Return a list of the message type of all topics in the bag"""
        return [ var[0] for var in self.bag.get_type_and_topic_info()[1].values() ]

    def start_time(self):
        """Return the timestamp of the first message in the bag"""
        return self.bag.get_start_time()

    def end_time(self):
        """Return the timestamp of the last message in the bag"""
        return self.bag.get_end_time()

    def data(self):
        """Return the data from all messages in the bag as a nested dict"""
        data_dict = {}
        for topic, msg, t in self.bag.read_messages():
            if not topic in data_dict:
                data_dict[self.topic_to_field(topic)] = self.initialize_topic(msg)
            self.extract_msg_data(msg, data_dict[self.topic_to_field(topic)])
        return data_dict

    def initialize_topic(self, msg):
        """Recursively intialize the data structure from a message"""
        if not hasattr(msg, '__slots__'):
            return []
        else:
            recurse_dict = {}
            for slot in getattr(msg, '__slots__'):
                recurse_dict[slot] = self.initialize_topic(getattr(msg, slot))
            return recurse_dict

    def extract_msg_data(self, msg, data):
        """Recursively extract data from a message"""
        if not hasattr(msg, '__slots__'):
            data.append(msg)
        else:
            for slot in getattr(msg, '__slots__'):
                self.extract_msg_data(getattr(msg, slot), data[slot])

    def topic_to_field(self, topic):
        """Convert a topic name to a valid variable name"""
        return topic[1:].replace('/','_')
